src/ControlSystem/Event.hpp

```diff
@@ -0,0 +1,125 @@
+/// will be created during control system initialization for each \ref
+/// control_system::protocols::Measurement "measurement".
+///
+/// These events must be added to the \ref
+/// Options::protocols::FactoryCreation "factory_creation" struct in
+/// the metavariables, even though they cannot be created from the
+/// input file.  The `control_system::control_system_events`
+/// metafunction provides the list of events to include.
+template <typename ControlSystems>
+class Event : public ::Event {
+  static_assert(tmpl::size<ControlSystems>::value > 0);
+  using measurement = typename tmpl::front<ControlSystems>::measurement;
+  static_assert(tmpl::all<ControlSystems,
+                          std::is_same<metafunctions::measurement<tmpl::_1>,
+                                       tmpl::pin<measurement>>>::value);
+  static_assert(
+      tt::assert_conforms_to<measurement, protocols::Measurement>);
+
+  template <typename ControlSystem>
+  using process_measurement_for_control_system =
+      typename ControlSystem::process_measurement;
+
+  using callbacks = tmpl::transform<
+      ControlSystems,
+      tmpl::bind<process_measurement_for_control_system, tmpl::_1>>;
+
+ public:
+  /// \cond
+  explicit Event(CkMigrateMessage* /*unused*/) noexcept {}
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(Event);  // NOLINT
+  /// \endcond
+
+  static constexpr bool factory_creatable = false;
+  Event() = default;
+
+  using argument_tags = tmpl::list<Tags::DataBox>;
+
+  template <typename DbTags, typename Metavariables, typename ArrayIndex,
+            typename Component>
+  void operator()(const db::DataBox<DbTags>& box,
+                  Parallel::GlobalCache<Metavariables>& cache,
+                  const ArrayIndex& array_index,
+                  const Component* const component) const noexcept {
+    const LinkedMessageId<double> measurement_id{
+        db::get<::Tags::Time>(box),
+        db::get<::evolution::Tags::PreviousTriggerTime>(box)};
+    tmpl::for_each<typename measurement::submeasurements>(
+        [&array_index, &box, &cache, &component,
+         &measurement_id](auto submeasurement) noexcept {
+          using Submeasurement = tmpl::type_from<decltype(submeasurement)>;
+          db::apply<Submeasurement>(box, measurement_id, cache, array_index,
+                                    component, ControlSystems{});
+        });
+  }
+
+  using is_ready_argument_tags = tmpl::list<::Tags::Time>;
+
+  template <typename Metavariables, typename ArrayIndex, typename Component>
+  bool is_ready(const double time, Parallel::GlobalCache<Metavariables>& cache,
+                const ArrayIndex& array_index,
+                const Component* const component) const noexcept {
+    // This checks all the functions of time, not just those the
+    // measurement is being made for.  We need all of them to access
+    // the evolved variables.
```

Comment:
Change self-request: We need these to access the coordinates.  The variables themselves actually don't require them.